### ISSUE TITLE
Update TEO kinematic model (table and diagram)

### ISSUE THREAD

**Issue by PeterBowman** [roboticslab-uc3m/teo-software-manual#40]
The orientation of the TCP is not correct, compare [rightArmKinematics.ini](https://github.com/roboticslab-uc3m/teo-configuration-files/blob/8ced4c99e4137a48aff87397c274943b0aef32cb/share/kinematics/rightArmKinematics.ini) (same for the left arm) and [dh-drawing.png](https://github.com/roboticslab-uc3m/teo-developer-manual/blob/dcab8e5983bccc29f77f5802ade3be3d5080b0a5/assets/dh-drawing.png). That picture was imported from an older version of http://robots.uc3m.es/index.php/File:DH.png and it doesn't reflect the actual behavior of the manipulators (regardless of whether the grippers have been rotated or not).

Also, the naming of some axes is wrong (check *x4* and *x5* on the right arm) or missing (left arm).

**Comment by jgvictores** [comment]
I'd consider this a bit of a duplicate of https://github.com/roboticslab-uc3m/teo-main/issues/38

Note that it points to https://github.com/roboticslab-uc3m/teo-main/issues/23 (where I mention to please follow the [workflow-for-assets](https://github.com/roboticslab-uc3m/developer-manual/issues/26))

**Comment by PeterBowman** [comment]
Please either update kinematics-related .inis at teo-configuration-files or fix the diagram here, this is a bug that shouldn't have anything to do with the physical changes described at https://github.com/roboticslab-uc3m/teo-main/issues/23. I feel that https://github.com/roboticslab-uc3m/teo-main/issues/38 had been opened as a tracking issue for double-checking things, whereas I want to stress here that a nontrivial problem was found and it should not be merely taken as implied in a (to some extent) unrelated repo for which I didn't even know such an issue existed.

Could we just replace the diagram with its latest revision from the wiki?

**Comment by jgvictores** [comment]
I totally agree we can maintain https://github.com/roboticslab-uc3m/teo-main/issues/38 as a tracking issue and solve parts. I've updated it's description to include the workflow explicitly (first edit and maintain matching [dh-drawing.svg](https://github.com/roboticslab-uc3m/teo-developer-manual/blob/master/assets/editable/dh-drawing.svg) and [dh-table.ods](https://github.com/roboticslab-uc3m/teo-developer-manual/blob/master/assets/editable/dh-table.ods), then do the rest).

> Could we just replace the diagram with its latest revision from the wiki?

Two concerns here:
1. I'd really like to have the correct `.svg` in the repo before simply modifying the `.png`. If the corresponding original `.svg` is not in GitHub, older versions can be found in http://robots.uc3m.es/svn/TEOrepo/modelling/diagrams/editable/
1. Please also check parameters regarding legs, I'm pretty sure @munozyanez and/or @rsantos88 did changes there that might not be in older versions.

**Comment by jgvictores** [comment]
Considering this blocked by https://github.com/roboticslab-uc3m/teo-main/issues/38 and related to https://github.com/roboticslab-uc3m/developer-manual/issues/26

**Comment by jgvictores** [comment]
Using this issue to track https://github.com/roboticslab-uc3m/teo-developer-manual updates now that we have closed the grand "Review and Unify Joint Directions of Rotation (https://github.com/roboticslab-uc3m/teo-main/issues/39)"!

**Comment by jgvictores** [comment]
- Trunk: no changes
![photo_2019-04-10_17-29-29](https://user-images.githubusercontent.com/4448588/55892359-81b7e980-5bb6-11e9-8ca6-94ab70ceae21.jpg)
![Screenshot from 2019-04-08 19-35-53](https://user-images.githubusercontent.com/4448588/55744620-a89ae200-5a35-11e9-958e-d05019d90142.png)

**Comment by jgvictores** [comment]
See https://github.com/roboticslab-uc3m/teo-main/issues/38#issuecomment-480930979 for use of https://github.com/arcoslab/roboview (on Trunk for now).

**Comment by jgvictores** [comment]
- Right Arm: some changes
![photo_2019-04-10_17-30-34](https://user-images.githubusercontent.com/4448588/55892509-b9269600-5bb6-11e9-97e6-580ff97a9425.jpg)
![Screenshot from 2019-04-10 17-57-31](https://user-images.githubusercontent.com/4448588/55894297-2a1b7d00-5bba-11e9-9952-7965b0fcf49c.png)

**Comment by jgvictores** [comment]
- Left Arm: same changes as for right arm, not copying images again

**Comment by jgvictores** [comment]
- Right Leg and Left Leg: some changes, both very similar except for sign of `l16` (`D` of link 4), note how the lower links hav a slight inwards tilt
![photo_2019-04-11_11-39-58](https://user-images.githubusercontent.com/4448588/55948382-ad87ad80-5c50-11e9-905b-5982febabccd.jpg)
![Screenshot from 2019-04-11 11-53-39](https://user-images.githubusercontent.com/4448588/55948418-ca23e580-5c50-11e9-887b-338a12eab7ca.png)

**Comment by jgvictores** [comment]
- Head: no changes
![photo_2019-04-11_12-07-00](https://user-images.githubusercontent.com/4448588/55949345-a82b6280-5c52-11e9-8e67-e86a5620d550.jpg)
![head](https://user-images.githubusercontent.com/4448588/55949350-a9f52600-5c52-11e9-90b6-6015264cef83.png)

**Comment by jgvictores** [comment]
- Transformations: adding new root to right/left arms, will also update old transformation from hip to arms
![photo_2019-04-11_12-38-11](https://user-images.githubusercontent.com/4448588/55951190-f04c8400-5c56-11e9-81d8-9e9e275352a5.jpg)
![Screenshot from 2019-04-11 12-32-03](https://user-images.githubusercontent.com/4448588/55951213-fe9aa000-5c56-11e9-8475-2f8c51b4a8b0.png)

**Comment by jgvictores** [comment]
Regarding https://github.com/roboticslab-uc3m/teo-developer-manual/commit/6ed9975cd4279ac5b0dafbc861ef71d0ada10f26, you can even stack them up, say `dh-root-rightArm.csv`+`dh-rightArm.csv` and `dh-root-leftArm.csv`+`dh-leftArm.csv` to make sure everything is correct from `root` to end-effector:

![bimanip](https://user-images.githubusercontent.com/4448588/55951734-39510800-5c58-11e9-8b4d-1a04ea6fd05a.png)

**Comment by jgvictores** [comment]
- Transformations: adding new root to right/left legs, will also update old transformation from root to legs (minor typo in pic: top `x_1` should be `x_0`)
![photo_2019-04-11_13-06-21](https://user-images.githubusercontent.com/4448588/55953076-74a10600-5c5b-11e9-811f-b124d6e4768c.jpg)

**Comment by jgvictores** [comment]
Regarding https://github.com/roboticslab-uc3m/teo-developer-manual/commit/b7cf504f8b21c9d0eeb3b0b69afccad0a0e062a5, again we can concatenate, say `dh-root-rightLeg.csv`+`dh-rightLeg.csv` and `dh-root-leftLeg.csv`+`dh-leftLeg.csv` to go "Live on two Legs":

![liveOnTwoLegs](https://user-images.githubusercontent.com/4448588/55957294-3fe67c00-5c66-11e9-82d9-2f9342be2327.png)

**Comment by jgvictores** [comment]
Due to what is illustrated in the pic, we are changing the head joint order (tracked at https://github.com/roboticslab-uc3m/teo-main/issues/48), to use the DH below:

![photo_2019-04-11_19-49-01](https://user-images.githubusercontent.com/4448588/55979630-f1030b80-5c92-11e9-87d3-27611f0f52d2.jpg)

**Comment by jgvictores** [comment]
> Due to what is illustrated in the pic, we are changing the head joint order (tracked at [roboticslab-uc3m/teo-main#48](https://github.com/roboticslab-uc3m/teo-main/issues/48)), to use the DH below:

False alarm, we can use a solution that involves the neck starting in a lower position as in the figure (closing [roboticslab-uc3m/teo-main#48](https://github.com/roboticslab-uc3m/teo-main/issues/48) and https://github.com/roboticslab-uc3m/teo-configuration-files/pull/19 as `invalid`):

UPDATE: blanks filled in at https://github.com/roboticslab-uc3m/teo-developer-manual/issues/40#issuecomment-482536458
![photo_2019-04-12_10-59-56](https://user-images.githubusercontent.com/4448588/56025328-36b2e900-5d12-11e9-9e57-f33e35875ba9.jpg)

**Comment by jgvictores** [comment]
- Head updated at https://github.com/roboticslab-uc3m/teo-developer-manual/commit/ad023dd74d1c24f9507005458a91620886c31e72 and https://github.com/roboticslab-uc3m/teo-developer-manual/commit/d5f15384e30a66fa5038a5f3313df03dcebaec91 is now taller because starts below:
![photo_2019-04-12_13-05-27](https://user-images.githubusercontent.com/4448588/56033044-ee9cc200-5d23-11e9-9ee9-e39cd7ee97d1.jpg)
![Screenshot from 2019-04-12 13-05-58](https://user-images.githubusercontent.com/4448588/56033047-f0668580-5d23-11e9-84d6-69886d5f4564.png)

**Comment by jgvictores** [comment]
Regarding these commits, again we can concatenate, `dh-root-head.csv`+`dh-head.csv` for the final trunk+head adopted solution:
![photo_2019-04-12_13-11-11](https://user-images.githubusercontent.com/4448588/56033458-f6a93180-5d24-11e9-9745-10ea33418d0f.jpg)
![Screenshot from 2019-04-12 13-13-35](https://user-images.githubusercontent.com/4448588/56033471-fad54f00-5d24-11e9-8dbd-0e3392a3c64d.png)

**Comment by jgvictores** [comment]
Big merge at https://github.com/roboticslab-uc3m/teo-developer-manual/commit/0c8c519c48edc89e06cc2212419661ad909f1bb7

- [x] Pending: render the new `csv/dh-root-*.csv` files in the gitbook.

**Comment by jgvictores** [comment]
> * [x]  Pending: render the new `csv/dh-root-*.csv` files in the gitbook.

Added at https://github.com/roboticslab-uc3m/teo-developer-manual/commit/13d36106fd6a9c45e986e7ba203520ee8c71151a

**Comment by jgvictores** [comment]
As a next step of this task, I'd focus on the end-effectors (kinematic chain tips):
- [x] `rightArm`/`leftArm`:
   - Reduce `l_8`, most probably to include the F/T sensor but not any hand
   - Add transformations from this end-effector to TCPs for different types of hands
- [x] `rightLeg`/`leftLeg`: Maybe we'd want some additional changes for walking. Maybe we do not even want root to foot but the opposite. Requires though that should be done at new issues: marking as "okay for now".
- [x] `head`: Transformations seem relatively okay for now, but may be subject to change in the future (depends on `vision`).

**Comment by rsantos88** [comment]
tested [trunkAndRightArmKinematics](https://github.com/roboticslab-uc3m/teo-configuration-files/commit/08407fa1ccdf29cb75d5a277d4bc373d125fd3f4) and [trunkAndLeftArmKinematics](https://github.com/roboticslab-uc3m/teo-configuration-files/commit/3d584e14e52f0e837c51ba79aad9c51d8a12c589) on teoSim and it works great!

**Comment by jgvictores** [comment]
In addition to the [comments above](https://github.com/roboticslab-uc3m/teo-developer-manual/issues/40#issuecomment-485465734), will also focus on scripts to generate `H0` and `HN` transformations. Note that `teo-configuration-files` say they use `asibot-main/example/cpp/testTinyMath.cpp`: https://github.com/roboticslab-uc3m/teo-configuration-files/blob/23703e1e29182a43c491efe04d4f5a9427d6008a/share/kinematics/leftArmKinematics.ini#L3

However:
- `asibot-main/example/cpp/testTinyMath.cpp` is now [roboticslab-uc3m/kinematics-dynamics/examples/cpp/exampleYarpTinyMath/exampleYarpTinyMath.cpp](https://github.com/roboticslab-uc3m/kinematics-dynamics/blob/530561b45c238a6f18cffa30d20d90bb25ac587c/examples/cpp/exampleYarpTinyMath/exampleYarpTinyMath.cpp)
- I'd prefer to use `PyKDL`

**Comment by jgvictores** [comment]
> However:
> 
> * `asibot-main/example/cpp/testTinyMath.cpp` is now [roboticslab-uc3m/kinematics-dynamics/examples/cpp/exampleYarpTinyMath/exampleYarpTinyMath.cpp](https://github.com/roboticslab-uc3m/kinematics-dynamics/blob/530561b45c238a6f18cffa30d20d90bb25ac587c/examples/cpp/exampleYarpTinyMath/exampleYarpTinyMath.cpp)
> * I'd prefer to use `PyKDL`

Opened https://github.com/roboticslab-uc3m/kinematics-dynamics/issues/177 for this task.

**Comment by jgvictores** [comment]
> `rightArm`/`leftArm`:
> * Reduce `l_8`, most probably to include the F/T sensor but not any hand
> * Add transformations from this end-effector to TCPs for different types of hands

Done via https://github.com/roboticslab-uc3m/teo-developer-manual/commit/4a2431a7804bacf2e0b378101a4a208c768e8012. Fetch now has Z pointing down, with old `a 2 cm debajo chapa final mano` via new `lf` (F/T included in `l8`).

Pending: update diagrams.

**Comment by jgvictores** [comment]
Done at https://github.com/roboticslab-uc3m/teo-developer-manual/commit/cabf02d5438e2735a4983b82f9ff0c6fca2fe6cd

**Comment by jgvictores** [comment]
Attaching DH coord systems for Fetch
![dh-fetch](https://user-images.githubusercontent.com/4448588/57242616-42a17a80-7034-11e9-8c91-66c9f10c5b68.jpg)
